KL Divergence Based Feature Selection

zohairshafi opened the pull request:
I’ve implemented a new way to rank feature importance by computing the Kullback–Leibler divergence between the distributions of each binary feature for the two classes (label = 1 versus label = 0). The idea is that a feature whose probability mass shifts noticeably when conditioned on one class instead of the other will be highly predictive, so it receives a higher importance score.

The additions conform to the API and project structure.

zohairshafi committed:
Create kl_divergence.py

Signed-off-by: Zohair Shafi <zohair.shafi@gmail.com>

zohairshafi committed:
Update selector.py

Signed-off-by: Zohair Shafi <zohair.shafi@gmail.com>

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+from feature.utils import Num, get_task_string, check_true
+
+
+from tqdm import tqdm
```
is this needed. Notice it is not currently part of reqs.txt? I would remove if not absolutely necessary

skadio commented on feature/selector.py:
```diff
@@ -166,6 +167,28 @@ def _validate(self):
 
+    class KL_Divergence(NamedTuple):
+        """
+
```
remove extra line

skadio commented on feature/selector.py:
```diff
@@ -166,6 +167,28 @@ def _validate(self):
+    class KL_Divergence(NamedTuple):
+        """
+
+        Computes the distribution of a given feature for instances where label == 1 and label == 0.
```
very nice explanation!

skadio commented on feature/selector.py:
cool to see how "easy" it is to introduce new method. The changes needed are very minimal

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+
+class _KL_Divergence(_BaseSupervisedSelector, _BaseDispatcher):
+
+    def __init__(self, seed: int, num_features: Num,):
```
remove dangling "," at the end

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+from tqdm import tqdm
+
+
+class _KL_Divergence(_BaseSupervisedSelector, _BaseDispatcher):
```
I don't think this needs to uptake Dispatcher. We typically use dispatcher when the method is an umbrella method (like Linear) and the exact implementer is decided eventually upon the parameters (Linear but Lasso etc.)

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+
+    def fit(self, data: pd.DataFrame, labels: pd.Series) -> NoReturn:
+
+        check_true(len(np.unique(labels)) == 2, TypeError("Only binary labels are supported for KL Divergence"))
```
do we typically do this check in fit()? How about other supervised methods, do they have this?

skadio commented on feature/selector.py:
```diff
@@ -166,6 +167,28 @@ def _validate(self):
+        Uses KL divergence between the two distributions as an importance score,
+        where a higher value indicates greater discriminative power of the feature
+        with respect to the binary label.
+
```
Do you want to add a comment on complexity of the method? For how many rows (observations) is this typically supposed to be run? OR when does it become a bottleneck, roughly speaking. I am looking for sth like, run this for less than 10K rows, or 100K rows Sth like that

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+        
+            kl_mat[i] = np.sum(kl)
+
+        self.abs_scores = kl_mat.flatten()
```
checking the math here.. kl_mat is guaranteed to be positive? notice that this is "absolute scores"

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+        
+        kl_mat = np.zeros((data.shape[1], 1))
+        data = data.values
+        label_categories = np.unique(labels)
```
you already did a np.unique() before, no need to pay the price again

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+
+        check_true(len(np.unique(labels)) == 2, TypeError("Only binary labels are supported for KL Divergence"))
+        
+        kl_mat = np.zeros((data.shape[1], 1))
```
IF you call data.shape[1] as `num_features` the code becomes more readable

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+import pandas as pd
+import numpy as np
+
+from scipy.special import kl_div, rel_entr
```
where is `kl_div` used?

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+
+from scipy.special import kl_div, rel_entr
+from feature.base import _BaseSupervisedSelector, _BaseDispatcher
+from feature.utils import Num, get_task_string, check_true
```
where is `get_task_string` is used? Check your PyCharm, and remove unused imports (or better, click on Optimize Imports)

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+        
+        for i in tqdm(range(data.shape[1])):
+            
+            pos_idx = np.where(labels == label_categories[0])[0]
```
how do you know label[0] is positive? Can you check if this could actually be the other way around?

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+        for i in tqdm(range(data.shape[1])):
+            
+            pos_idx = np.where(labels == label_categories[0])[0]
+            neg_idx = np.where(labels == label_categories[1])[0]
```
also, something is off in this loop. we are getting positive negative labels repeatedly for each feature. Cannot you grab this once outside the loop and use it, sth like 

```
pos_idx = labels[labels == 1].index
neg_idx = labels[labels == 0].index
```

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+            pos_idx = np.where(labels == label_categories[0])[0]
+            neg_idx = np.where(labels == label_categories[1])[0]
+            
+            f1 = np.histogram(data[pos_idx, i], bins = 100)[0]
```
make 100 a parameter? Should we push that to main API, and default to 100. So KLD will have (num_features, num_bins) as params?

skadio reviewed (commented):
@zohairshafi 

Thanks for contributing a net-new method to Selective library. I left some (minor) code comments regarding the implementation of the method. Should be easy to fix

However, on a second look, I now realize that I misguided you. My apologies. We already have "Statistical" methods so actually kl_divergence should have been a method of that umbreall. Your existing implementation should work as-is but it would be great to push this under Statistical please. 

Beyond those, there are also some "library" level things that must happen. 
* Make sure reqs.txt is still good 
* Under tests/ folder add test_kl_divergence.py. Skim the other tests to get a sense of what/how to test. Make sure to tests for all invalid scenarios as well (basically those you guard against setting features, or things like when someone runs this for unsupervised setting) 
* Please confirm that all existing tests + plus the new kl_divergence tests pass
* Update the Readme. 
  * Add the KLDivergence row in available methods, inside Statistical row
  * Add KLD method in the Benchmarking example 
* Increment the version in _version.py to 1.2.0 
* In the Changelog add a row that KLDivergence is now added in this version. add sth like `thanks to @zohairshafi for contributing this method`.

zohairshafi commented:
Sounds good, I will change the structure to make it part of the Statistical framework as well as make the code changes before committing again.

takojunior commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,69 @@
+            f1 = np.histogram(data[pos_idx, i], bins = 100)[0]
+            f2 = np.histogram(data[neg_idx, i], bins = 100)[0]
+        
+            f1 = f1 / np.sum(f1)
```
To check my understanding, this is normalization? to make f1, f2 in a range [0, 1]? So these become 2 distributions, one for positive and one for negative, and then get kl between the two?

zohairshafi replied:
Yes this is correct!

skadio replied:
consider adding this explanation as code comment. this info would be lost after the PR

zohairshafi replied:
Done!

takojunior commented on feature/selector.py:
```diff
@@ -26,6 +26,7 @@
 from feature.base import _BaseDispatcher, _BaseSupervisedSelector, _BaseUnsupervisedSelector
 from feature.correlation import _Correlation
 from feature.linear import _Linear
+from feature.kl_divergence import _KL_Divergence
```
Cool

takojunior commented on feature/selector.py:
```diff
@@ -166,6 +167,28 @@ def _validate(self):
+        num_features: Num = 0.0
+
+        def _validate(self):
+            check_true(isinstance(self.num_features, (int, float)), TypeError("Num features must a number."))
```
Typo: ...must be a number.

Commit:
Updates

zohairshafi commented:
Changes pushed in latest commit: 

1. _BaseDispatcher removed from KL Divergence class. 
2. KL Divergence is now part of statistical.py
3. check_true in fit() function -- this needs to be checked only when data is actually passed in. The first time that happens is in the fit() function. 
4. Do you want to add a comment on complexity of the method? - This is using the numpy histogram and spicy relative entropy methods - I don't imagine they should have an issue with extremely large datasets. Nothing I've seen so far. 
5. kl_mat is guaranteed to be positive? - Yes, I have also changed the code a little bit to ensure we calculate KL divergence in both directions since it is not symmetric. 
6. how do you know label[0] is positive? Can you check if this could actually be the other way around? - This works both ways - it can be any two unique class labels, not necessarily 0/1. The pos_idx and neg_idx names are just for readability. I changed the variable names to class_one_idx and class_two_idx now. 
7. KLD will have (num_features, num_bins) as params? - This made sense when KLD was its own class - but as part of the Statistical class, I'm not sure how to add an extra input parameter without breaking other code. Could you advise please? 
8. Other changes to variable names / efficiency based changes / unused imports etc.. have been done. 
9. Readme has been updated. The hyperlink to KL Divergence points to the wikipedia article. Is that okay? 
10. Version has been updated. Changelog has been updated. Requirements does not need updating. 

To Do: 
1. Add tests for KLD. 
2. Run and ensure all tests work.

Question: 
1. Do the imports need to be updated from "from feature import ..." to "from selective.feature import ..."? I keep getting the error "ModuleNotFoundError: No module named 'feature'" unless I add selective/ to the system path. I think this problem goes away in a pip install though.

skadio commented on README.md:
```diff
@@ -48,7 +48,7 @@ print("Scores:", list(selector.get_absolute_scores()))
 | [Variance per Feature](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html) |                                                                                                                                                                                                                                                                                  ...
 |   [Correlation pairwise Features](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.corr.html)   |                                                                                                                                                                                                                                                                     [Pearson Cor ...
-|    [Statistical Analysis](https://scikit-learn.org/stable/modules/feature_selection.html#univariate-feature-selection)     |                                                                                                             [ANOVA F-test Classification](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html) <br> [F-value Regression](https://scikit-l ...
+|    [Statistical Analysis](https://scikit-learn.org/stable/modules/feature_selection.html#univariate-feature-selection)     |                                                                                                             [ANOVA F-test Classification](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html) <br> [F-value Regression](https://scikit-l ...
```
is this lex-ordered?

skadio commented on README.md:
```diff
@@ -82,6 +82,7 @@ selectors = {
   "stat_anova": SelectionMethod.Statistical(num_features, method="anova"),
   "stat_chi_square": SelectionMethod.Statistical(num_features, method="chi_square"),
   "stat_mutual_info": SelectionMethod.Statistical(num_features, method="mutual_info"),
+  "stat_kl_divergence": SelectionMethod.Statistical(num_features, method="kl_divergence"),
```
nice!

skadio replied:
(just lex-order it)

skadio commented on CHANGELOG.txt:
```diff
@@ -2,6 +2,12 @@
+August 7, 2025 1.2.0
+-------------------------------------------------------------------------------
+
+- Added KL Divergence based feature selection for binary labels. Thanks to @zohairshafi for contributing this method. 
```
cool!

skadio commented on feature/statistical.py:
```diff
@@ -41,7 +43,8 @@ def __init__(self, seed: int, num_features: Num, method: str):
                         # "classification_maximal_info": MINE(), # dropped
-                        "unsupervised_variance_inflation": variance_inflation_factor}
+                        "unsupervised_variance_inflation": variance_inflation_factor, 
+                        "kl_divergence" : _KL_Divergence(num_features = self.num_features, seed = self.seed)}
```
make this lex-order

skadio commented on feature/statistical.py:
```diff
@@ -56,6 +59,8 @@ def dispatch_model(self, labels: pd.Series, *args):
         # Get statistical scoring function
         if method == "variance_inflation":
             score_func = self.factory.get("unsupervised_" + method)
+        elif method == "kl_divergence":
```
lex order

skadio commented on feature/statistical.py:
```diff
@@ -82,6 +89,7 @@ def fit(self, data: pd.DataFrame, labels: pd.Series) -> NoReturn:
         if self.method == "variance_inflation":
             # VIF is unsupervised, regression between data and each feature
             self.abs_scores = np.array([variance_inflation_factor(data.values, i) for i in range(data.shape[1])])
+
```
remove extra line

skadio commented on feature/selector.py:
```diff
@@ -227,7 +234,7 @@ def _validate(self):
                 check_true(self.num_features <= 1, ValueError("Num features ratio must be between [0..1]."))
             # "maximal_info" dropped
-            check_true(self.method in ["anova", "chi_square", "mutual_info", "variance_inflation"],
+            check_true(self.method in ["anova", "chi_square", "mutual_info", "variance_inflation", "kl_divergence"],
```
lex

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,59 @@
+        
+        kl_mat = np.zeros((self.num_features, 1))
+        X = X.values
+        label_categories = np.unique(y)
```
move this before check_true to avoid np.unique() cost. labels column can be large

skadio commented on feature/kl_divergence.py:
```diff
@@ -0,0 +1,59 @@
+
+class _KL_Divergence(_BaseSupervisedSelector):
+
+    def __init__(self, seed: int, num_features: Num, num_bins: Num = 100):
```
agree keeping num_bins here.. not obvious to me how to make this part of Statistical without kwwargs which would require changing a lot of code.. this is good as-is

skadio reviewed (commented):
All changes look good to me. Once you add KLD tests and confirm all tests pass, we can merge.

I am not sure if we directly merge this .. OR .. "we" copy your fork internally and merge from here.. I don't know.. does not matter for now. Let's just get the tests going and we can decide how to merge later.

takojunior commented on feature/_version.py:
```diff
@@ -2,4 +2,4 @@
 
-__version__ = "1.1.2"
\ No newline at end of file
+__version__ = "1.2.0"
```
Cool. I am a maintainer of selective in pypi. I can upload the new version to pypi @skadio . And I see we should also create a new release on github.

Commit:
Updates

zohairshafi commented:
I've made another commit with the changes requested. 
In terms of tests, I'm a little confused - in test_benchmark.py, where are the hard coded numbers coming from?

skadio commented on tests/test_stat_kl.py:
```diff
@@ -0,0 +1,79 @@
+from tests.test_base import BaseTest
+
+
+class TestKL(BaseTest):
```
nice!!

skadio reviewed (commented):
@zohairshafi this looks good to go (just resolve comments to see if there is anything else dangling, I don't see anything).

Re: Tests, can you confirm that 
1) the new kl_tests pas? 
2) the existising tests pass? 

I have a feeling that 1 is OK but 2 is not OK, yes? 

That could most likely be because of reqs.txt does not add exact version, so in your new environment, the floating outputs don't really match anymore to hardcoded values, is this what you see? Does that only happen in test_benchmaks, or what else is failing? 

I think if the above is the only thing for 2 not passing, that's probably OK. 

Once you confirm, we can consider merging.

@takojunior to merge, "we" might need to create a fork within Selective, copy the files changed here and then press "merge" from our side. (as opposed to accepting a merge from non-company account now that @zohairshafi's internship here is concluded)

zohairshafi committed:
KL Test Updates

Commit:
Comments

zohairshafi commented:
KL Tests are done. Other tests with hardcoded values pass, but there are some issues with LightGBM. I wonder if this is also related to the version issues. Also possibly related to the version issues, there is an error I see with respect to Textwiser -  initially I thought it was a Mac specific error but now when I try to run it on Linux, I get the same error: 

```
RuntimeError: Failed to import transformers.models.layoutlmv2.feature_extraction_layoutlmv2 because of the following error (look up to see its traceback):
partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)

```

I think everything else should also be done now!

skadio commented:
> KL Tests are done. Other tests with hardcoded values pass, but there are some issues with LightGBM. I wonder if this is also related to the version issues. Also possibly related to the version issues, there is an error I see with respect to Textwiser - initially I thought it was a Mac specific error but now when I try to run it on Linux, I get the same error:
> 
> ```
> RuntimeError: Failed to import transformers.models.layoutlmv2.feature_extraction_layoutlmv2 because of the following error (look up to see its traceback):
> partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)
> ```
> 
> I think everything else should also be done now!

Sounds good! Can you go through unresolved comments to make sure all is addressed please?

@kuppulur do you recognize the TextWiser error? Is there an easy fix, or?

zohairshafi commented:
Yes, done! 
All comments had been resolved in the code previously - I went and resolved all comments in this thread as well.

kuppulur commented:
> > KL Tests are done. Other tests with hardcoded values pass, but there are some issues with LightGBM. I wonder if this is also related to the version issues. Also possibly related to the version issues, there is an error I see with respect to Textwiser - initially I thought it was a Mac specific error but now when I try to run it on Linux, I get the same error:
> > ```
> > RuntimeError: Failed to import transformers.models.layoutlmv2.feature_extraction_layoutlmv2 because of the following error (look up to see its traceback):
> > partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think everything else should also be done now!
> 
> Sounds good! Can you go through unresolved comments to make sure all is addressed please?
> 
> @kuppulur do you recognize the TextWiser error? Is there an easy fix, or?

Can @zohairshafi you post the full error please, I have never seen this error and I don't see any build failures on textwiser too (CI is green)

AshishPvjs commented:
@zohairshafi I also see that one of the tests in KL Divergence is failing. I have posted the traceback here can you take a look at it?

FAIL: test_kl_classif_top_k (test_stat_kl.TestKL)
Traceback (most recent call last):
File "/Users/*****/selective/tests/test_stat_kl.py", line 45, in test_kl_classif_top_k
self.assertListEqual(list(subset.columns), ['sepal length (cm)', 'sepal width (cm)'])
AssertionError: Lists differ: ['sepal length (cm)', 'petal length (cm)'] != ['sepal length (cm)', 'sepal width (cm)']

First differing element 1:
'petal length (cm)'
'sepal width (cm)'

['sepal length (cm)', 'petal length (cm)']
? -- ^^^^
['sepal length (cm)', 'sepal width (cm)']
? ++ ^^^

zohairshafi commented:
Hello, sorry I was a little caught up with school work - is there anything I can do now or all good?

takojunior commented:
> Hello, sorry I was a little caught up with school work - is there anything I can do now or all good?

Hi @zohairshafi, @AshishPvjs was able to push a fix to the test mentioned above. The expected value is ['sepal length (cm)', 'petal length (cm)'] for the test. Now it is all good.

zohairshafi commented:
Ahhh got it, thank you so much!
